File: 05_Linked List/M237. Delete Node in a Linked List.java
// Delete Node in a Linked List
// https://leetcode.com/problems/delete-node-in-a-linked-list

class Solution {
    public static class ListNode{
        int val;
        ListNode next;
        ListNode (int x){ val = x;}
    }

    public void deleteNode(ListNode node){
        node.val = node.next.val;
        node.next = node.next.next;

        // No need to explicitly set node.next.next = null as it will be garbage collected

    }

    public static void main(String[] args){
        Solution solution = new Solution();

        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);

        solution.deleteNode(head.next.next);

        ListNode curr = head;
        while (curr != null) {
            System.out.print(curr.val + " ");
            curr = curr.next;
        }

    }
}

File: JavaScript/05_Linked List/M237. Delete Node in a Linked List.js
// 237. Delete Node in a Linked List
// https://leetcode.com/problems/delete-node-in-a-linked-list/

// Definition for singly-linked list.
/**
 * Creates a list node
 */
function createListNode(val = 0, next = null) {
    return { val, next };
}

/**
 * @param {ListNode} node
 * @return {void} Do not return anything, modify node in-place instead.
 */
function deleteNode(node) {
    node.val = node.next.val;
    node.next = node.next.next;
}

// Helper function to print list
function printList(head) {
    const result = [];
    while (head !== null) {
        result.push(head.val);
        head = head.next;
    }
    return result;
}

const head = createListNode(1, createListNode(2, createListNode(3, createListNode(4, createListNode(5)))));
deleteNode(head.next.next); // Delete node with value 3
console.log(printList(head)); // Output: [1, 2, 4, 5]
